File: web/docs-site/examples/python/ex04_at_frames.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["laterite==0.10.1"]
# ///
"""Docs example — run it with `uv run ex04_at_frames.py`, from anywhere.

Everything above the `[start:code]` marker is machinery the page does not
show: the PEP 723 header that makes the file self-installing,
and the fixture arm that makes its repo-relative path resolve outside a
checkout.
"""

import urllib.request
from pathlib import Path

_FIXTURE = Path("examples/sample_site.ags")
_RAW = "https://raw.githubusercontent.com/niko86/laterite/main/examples/sample_site.ags"
if not _FIXTURE.exists():
    # Cold only for a reader running this outside the repo: in a checkout (and in
    # CI, cwd = repo root) the file is already there and this arm never executes,
    # so the gates stay offline. Fetching it — rather than rewriting the example
    # to an absolute path — is what keeps the text on the page the text you would
    # actually type.
    _FIXTURE.parent.mkdir(parents=True, exist_ok=True)
    _FIXTURE.write_bytes(urllib.request.urlopen(_RAW, timeout=30).read())

# --8<-- [start:code]
# what this shows: materialise a borehole's full record set as a dict of polars frames via .at(..).frames()
import laterite

ags = laterite.read("examples/sample_site.ags")
frames = ags.at("LOCA", ["BH01"]).frames()

# frames is a dict {group_code: polars frame}; pull one out by code (NOT q["SAMP"])
print(sorted(frames))
print(frames["SAMP"].height)

assert isinstance(frames, dict) and "SAMP" in frames and frames["SAMP"].height >= 1
# --8<-- [end:code]

File: web/docs-site/examples/node/ex04_at_frames.mjs
// what this shows: at(code, ids) fans out to a borehole's related groups; frames() materialises the record set.
// Needs the optional @duckdb/node-api peer (npm i @duckdb/node-api) — read/validate/fix don't.
import { read } from "laterite";
import assert from "node:assert/strict";

const ags = read("examples/sample_site.ags");

// at("LOCA", ids) walks the dictionary's parent graph down from LOCA and keeps
// only groups that carry rows for those locations; `.groups` is the manifest.
const q = ags.at("LOCA", ["BH01"]);
console.log(q.groups.sort());

// frames() returns { group: rows } for the whole related record set, each
// row-filtered to just the boreholes you asked for.
const frames = await q.frames();
console.log(Object.keys(frames).sort());
console.log(frames.SAMP.length);
ags.close();

assert.ok(q.groups.includes("LOCA") && q.groups.includes("SAMP"));
assert.ok(Array.isArray(frames.SAMP) && frames.SAMP.length >= 1);
